Add replace & replace! for collections

rfourquet opened the pull request:
This adds methods to `replace` for `AbstractArray`, `AbstractSet` and `Associative` and a corresponding inplace `replace!` function.
Did I miss already existing similar functionality in Base?

The order in which to put the 2 function arguments `pred` and `f` can be discussed.

Also, `replace!(f::Function, A, old)` is not included as it would be ambiguous with `replace!(pred, A, new)` (and a new function name would be necessary, e.g. `replaceif!`); But it is assumed that in most of the cases, the replacement value `f(x)` can be computed by `f(old)` and then use `replace!(A, old, f(old))`. Otherwise, `replace!(pred, f, A)` can be used.

rfourquet committed:
add replace & replace! for collections

tkelman commented on test/algorithms.jl:
```diff
@@ -0,0 +1,28 @@
+
+@testset "replace! & replace" begin
```
new test files would need to be added to test/choosetests.jl to actually run, but aren't there existing files with related tests this would fit with?

tkelman commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+module Algorithms
```
while reorganizing some of base into modules is a good idea, I don't see why these two functions make sense as the only thing in an Algorithms module

rfourquet replied:
Yes, I forgot to mention that in the OP. It's just that I didn't know to put those functions, but I open to suggestions.

ararslan replied:
Seems they'd simply belong in the files corresponding to the types they're implemented for? That is, base/abstractarray.jl, base/set.jl, and base/associative.jl. I agree it doesn't make sense to separate these into their own module.

rfourquet replied:
But where to put the general methods which work on `Union{AbstractArray,Associative,AbstractSet}` ?

nalimilan replied:
`find*` methods are defined in `array.jl` even when they apply to any iterator, which isn't ideal but I guess OK.

rfourquet replied:
I just saw that the `unique` methods are defined in `set.jl` and `multidimensional`; the reason seems to be that they are defined right after the needed tools (e.g. `Set` or multidim tools) are defined. This would make much more sense to me put those functions also in an "algorithm.jl" file. E.g. I would expect to find in "set.jl" only the functions implementing the interface of `Set`, and possibly methods specialized for `Set`, but functions merely using a set in their implementation have nothing to do there.

nalimilan replied:
But then what's are the boundaries of this file? Any function uses an "algorithm". Maybe we'd need a collections.jl file with all generic functions which apply to any collection. Anyway, better use an existing file for this PR, and open another one to regroup functions in a subsequent PR.

rfourquet replied:
I agree "algorithm" is a bit vague (I "algorihtms" like in the c++ STL...), but no problem I will update and follow existing practices here.

tkelman commented on base/exports.jl:
```diff
@@ -1351,6 +1351,9 @@ export
     nnz,
 
+# Algorithms module re-exports
+    replace!,
```
needs to be added to stdlib doc index

rfourquet replied:
Will do.

tkelman commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+julia> replace!(x->x.first=>3, Dict(1=>2, 3=>4), 1) do (k, v)
+           v < 3
+       end
+
```
there shouldn't be an empty line between input and output, should there?

rfourquet replied:
No indeed.

tkelman commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+
+Return a copy of collection `A` where all occurrences `x` for which
+`pred(x)` is true are replaced by `new` or `f(x)` (exactly one among
+`f` and `new` must be specified).
```
I think it would be better to just list the different signatures instead of trying to explain constraints on non-syntactic square brackets here

rfourquet replied:
Shall I put the 2 stacked signatures (one below the other), followed by a common doctstring?

tkelman replied:
I think that would be clearer, yes

iamed2 commented:
Given this is new behaviour not needed by Base itself, could it live in https://github.com/JuliaCollections/DataStructures.jl or another package instead?

rfourquet commented:
Why not, but this is very basic functionality... For example, C++ which is quite barebones in its standard library, has a `replace` function.

iamed2 commented:
A lot of what is in the C++ standard library (e.g. deque, sorted sets/maps, etc.) lives in other packages such as DataStructures.jl. Some functionality which exists in the C++ standard library was moved from Base out to other packages (e.g. priority queues, combinatorics, etc.). Code in packages can be updated more frequently and doesn't contribute to Base size or the time it takes to run Julia's tests.

Just thought I'd suggest it, since this has come up for discussion before in cases where the code is not as easily decoupled from Base.

nalimilan commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+Replace all occurrences `x` in collection `A` for which `pred(x)` is true
+by `new` or `f(x)` (exactly one among `f` and `new` must be specified).
+If `count` is specified, then replace at most `count` occurrences.
+This is the in-place version of [`replace`](@ref).
```
I don't think we usually refer to copying versions of the functions.

rfourquet replied:
I thougth it was a good thing to document more related functions in docstrings.

nalimilan commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+!!! note
+When `A` is an `Associative` or `AbstractSet` collection, if
+there are collisions among old and newly created keys, the result
+can be unexpected:
```
"can be unexpected" is a bit vague: the last insert key wins, which can be undefined results when the iteration order is undefined. For `Set`, is the order defined or not?

ararslan replied:
Also, I believe the text in Documenter `!!! note` blocks has to be indented four spaces.

rfourquet replied:
The order is not defined for `Set`, but reproducible, so I chose values which work to illustrate my point. Will try to find a better wording.

rfourquet replied:
Is it better now?

nalimilan commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+`pred(x)` is true are replaced by `new` or `f(x)` (exactly one among
+`f` and `new` must be specified).
+If `count` is given, then replace at most `count` occurrences.
+See the in-place version [`replace!`](@ref) for examples.
```
Better repeat examples (that's what we do e.g. for `sum`).

nalimilan commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+replace!(pred, A::ReplaceCollection, new, n::Integer=-1) = replace!(pred, y->new, A, n)
+
+"""
+    replace(pred, [f::Function], A, [new], [count])
```
`count` is called `n` for the `AbstractString` method. Better stay consistent (in one direction or the other).

nalimilan commented on base/algorithms.jl:
````diff
@@ -0,0 +1,119 @@
+```jldoctest
+julia> a = [1, 2, 3, 1];
+
+julia> replace!(isodd, a, 0, 2); a
````
Why `; a`? Doesn't the function return it already? BTW, maybe it would be useful to return `count` instead?

rfourquet replied:
It was to clarify that `a` is modified in-place, but this may not be necessary. As to return `count`, I see that it could be useful, but I'm uncomfortable with the inconsistency of having the non-mutating `replace` returning the collection instead.

nalimilan replied:
The inconsistency with `replace` doesn't sound a big issue to me, but I would be more concerned about consistency with other in-place functions.

nalimilan commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+If `count` is given, then replace at most `count` occurrences.
+See the in-place version [`replace!`](@ref) for examples.
+"""
+replace(pred, new::Function, A::ReplaceCollection, n::Integer=-1) = replace!(pred, new, copy(A), n)
```
Wrap line at 92 chars.

nalimilan commented on base/algorithms.jl:
````diff
@@ -0,0 +1,119 @@
+```
+"""
+function replace!(pred, new::Function, A::AbstractArray, n::Integer=-1)
+    n == 0 && return A
````
The `AbstractString` method uses `n=0` by default, so it means "unlimited". Your choice sounds better since it allows using `0` for a no-op, which can be useful if that's the result of a computation, but better adapt the existing method then. Would also make sense to check that `n > -1` to prevent unexpected things.

rfourquet replied:
Yes I try propose to update the existing method in #22325. I will document the behavior for negatibe inputs, but I'm not sure it's useful to prevent `n < -1`.

rfourquet replied:
I'm converting to a default value of `typemax(Int)`, like for the string method.

ararslan commented:
Technically since `replace` is defined in Base and this is extending it with Base types, having it in a package would be committing type piracy. Also, FWIW, I've found myself wanting functionality like this often.

nalimilan commented on base/algorithms.jl:
```diff
@@ -0,0 +1,119 @@
+    A
+end
+
+const ReplaceCollection = Union{AbstractArray,Associative,AbstractSet}
```
Maybe you should leave the signature unrestricted until we have a `MutableCollection` type? Else, it won't be extensible for custom package types.

rfourquet replied:
To avoid ambiguities, then `pred` has to be constrained to be a `Function` (or `Base.Callable`), which seems OK.

nalimilan commented on base/algorithms.jl:
````diff
@@ -0,0 +1,119 @@
+
+# Examples
+```jldoctest
+julia> replace!(Set([1, 2, 3]), 1, 0)
````
Why not illustrate this with a plain `Array` first?

rfourquet replied:
Because a previous method of the same method (hence in the same docstring) was already having an example with an `Array`, so I tried to vary.

nalimilan replied:
It's fine to show examples using more complex structures, but I would start with the simplest example. Else it's not immediately obvious whether `Set` is particularly associated with that method or whether that's an arbitrary variation.

nalimilan reviewed (commented):
Sounds useful, thanks!

nalimilan commented:
FWIW, I have found out for [`CategoricalArrays.recode`](https://github.com/JuliaData/CategoricalArrays.jl/pull/67/) that pairs offer a simple, flexible and efficient API for this kind of operation. For example, `recode(x, 1=>2, 3=>4)` can perform two replacement operations and the generated code is really fast (i.e. the loop over the pairs is completely unrolled).

Maybe something to consider to make `replace` more powerful while keeping the interface relatively simple. Also the difference between `Pair` and `Function` could help dispatch when it's not clear what's the meaning of an argument.

iamed2 commented:
> Technically since replace is defined in Base and this is extending it with Base types, having it in a package would be committing type piracy.
The string `replace` doesn't obey the same conventions (replaces subsets, not elements) and could actually cause confusion or method conflicts if there was an iterator version of `replace` defined (and I'll probably write one if it doesn't make it into this PR).

rfourquet commented:
>  `recode(x, 1=>2, 3=>4)` can perform two replacement operations and the generated code is really fast 

I don't know CategoricalArrays yet, but for the `replace!(a, old, new)` method can be optimized greatly for `Associative` and `AbstractSet` (or sorted arrays).
 
> The string replace doesn't obey the same conventions (replaces subsets, not elements) 

It may be possible to replace subsets for collection too... (but than may require to be more strict type for the `old` value for single element replacement, e.g. `typeof(A)`)

nalimilan commented:
> I don't know CategoricalArrays yet, but for the replace!(a, old, new) method can be optimized greatly for Associative and AbstractSet (or sorted arrays).

Sorry, I guess my comment wasn't clear. What I mean is that using pairs allows `recode(x, 1=>2, 3=>4)` to do the equivalent of `replace(replace(x, 1, 2), 3, 4)`, but in a single pass, and with quite efficient code. This has nothing categorical-specific.

The element vs. subset debate reminds me of issues discussed in [the Search & Find Julep](https://github.com/JuliaLang/Juleps/blob/master/Find.md), with `findfirst` vs. (the proposed) `findseq`. But I don't really see the problem with strings: there is no ambiguity since `replace("qwerty" ,"qw", "")` isn't ambiguous. As @rfourquet it could be made to work for general arrays with some complications; might be worth using a separate `replaceseq` function for that. But it's so simple and natural for strings that it would be too bad not to support that feature.

iamed2 commented:
Seems like that concern does fit in with the Julep. If there was a `replaceseq` function, the String version would definitely fit better there, though the naming suffers. 

Lately I've been feeling that the string methods don't really jive with their iterator/array counterparts. But that's a general issue for somewhere else. 

> But I don't really see the problem with strings: there is no ambiguity since replace("qwerty" ,"qw", "") isn't ambiguous.

Suppose there was an iterator version of replace `replace{T}(itr{T}, old, new) -> ReplaceIterator{eltype(T)}` and `replace{T}(pred, itr{T}, new) -> ReplaceIterator{eltype{T}}`. If the first format is called on a String, the existing String method would take precedence. But if the user changed their call to call to the second format, the iterator version would be called and they would get a `ReplaceIterator{Char}` instead.

nalimilan commented:
> Suppose there was an iterator version of replace replace{T}(itr{T}, old, new) -> ReplaceIterator{eltype(T)} and replace{T}(pred, itr{T}, new) -> ReplaceIterator{eltype{T}}. If the first format is called on a String, the existing String method would take precedence. But if the user changed their call to call to the second format, the iterator version would be called and they would get a ReplaceIterator{Char} instead.

Yes, but at that point we would also add a specialized `String` method for the second form, returning a string. Anyway, `replace` isn't going to start returning an iterator, that would be `Base.Iterators.replace`.

iamed2 commented:
> Anyway, replace isn't going to start returning an iterator, that would be Base.Iterators.replace.

Ah, right, I forgot about that deprecation, sorry.

rfourquet commented:
I think I addressed all the comments. I still have to know how to set the link to `replace` from the doc of `replace!` point to the collection version, not to the string version; and to not have a duplicate docstring for the collection version...

On the design level:

1) I think @nalimilan's example of using `replace(A, old=>new)` is interesting; but Allowing multiple pairs would not be compatible with a "count" argument (`n`): `replace(A, p::Pair..., n=typemax(Int))` is not valid... As using a `Dict` of those pairs internally would probably be necessary for efficiency, it could be required to pass a `Dict` directly instead: `replace(A, repl::Dict, n=typemax(Int))`. 

2) there are two forms when a predicate is given:
```
replace(pred, A, new, [n])
replace(pred, f::Callable, A, [n])
```
It would be cleaner to have instead `replace(pred, A, f::Callable, [n])` for the second, to keep the same argument order, but this would collide when `A` can store callables... Do we care enough for this case? (I guess yes)

nalimilan commented:
> I think I addressed all the comments. I still have to know how to set the link to replace from the doc of replace! point to the collection version, not to the string version; and to not have a duplicate docstring for the collection version...

Can you elaborate? I think generally we point the docstring to the *function*, which concatenates docstrings for all methods. That's not ideal, but at least it includes the expected docstring.

> On the design level:
> 
>     I think @nalimilan's example of using replace(A, old=>new) is interesting; but Allowing multiple pairs would not be compatible with a "count" argument (n): replace(A, p::Pair..., n=typemax(Int)) is not valid... As using a Dict of those pairs internally would probably be necessary for efficiency, it could be required to pass a Dict directly instead: replace(A, repl::Dict, n=typemax(Int)).

Actually, no, the pairs should be passed as a varargs so that the compiler is able to unroll the checks rather than using a loop. With a small number of pairs, a dict isn't faster than repeated equality tests.

But you're right that the handling of `n` would be problematic in that case. We could make it a keyword argument. I guess it really depends on whether we want to make `replace` very general or rather keep it as simple as possible.

>     there are two forms when a predicate is given:
> 
> replace(pred, A, new, [n])
> replace(pred, f::Callable, A, [n])
> 
> It would be cleaner to have instead replace(pred, A, f::Callable, [n]) for the second, to keep the same argument order, but this would collide when A can store callables... Do we care enough for this case? (I guess yes)

Yes, that's a tough call. The API based on pairs would solve this.

nalimilan commented on base/array.jl:
```diff
@@ -1987,3 +1987,139 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+
+## replace/replace! ##
+
+# NOTE: to implement replace! and replace for a new type T, it's enough to define:
```
I don't think this is the place to document this. Though I don't really know where it could go. Also, I don't like recommending people to override an unexported function whose name starts with an underscore. Really, they can reimplement the three lines in `replace` that appear before the call to `_replace`, that's a negligible repetition.

rfourquet replied:
Ok, I will replace `_replace` by `replace` in the recommandation. Until a better place is suggested, I will keep it here though.

nalimilan replied:
I'd rather not have this at all, AFAIK we don't generally do this, and people are not supposed to read this file to find out...

rfourquet replied:
This note is intended also for people adding a new method in Base... What do you think about keeping/removing this note @tkelman ?

rfourquet commented:
> Can you elaborate? I think generally we point the docstring to the function, which concatenates docstrings for all methods. That's not ideal, but at least it includes the expected docstring.

By specifying the signatures in .md files, it's possible to show only the docstring for the string version in "strings.md", and the general collections version in "collection.md". But unfortunately, the link from `replace!` choses to point at the string version. In the docs of Documenter, I didn't see a way to address that, someone has an idea? otherwise I will simply remove the link. 

> Actually, no, the pairs should be passed as a varargs so that the compiler is able to unroll the checks rather than using a loop. With a small number of pairs, a dict isn't faster than repeated equality tests.

OK, I see. Then indeed, why not set the `count` argument as a keyword. It may be clearer actually (and I expect it not to be used so often), and I understood that keyword args are meant to be faster eventually. 

> The API based on pairs would solve this.

I don't see that. Assuming there is `replace(A, oldnew::Pair...; [count])`, how do you propose the version with a predicate to be? (for which `old` is not necessary anymore).
One possibility is to remove `replace(pred, A, new)`, and to merge merge `f` and `pred` for the last version, which can be more powerful:
`replace(pred_f, A)`, with `pred_f` returning `Nullable` for the replacement value. This would allow to implement the pair version in terms of this `pred_f` (finding whether a pair matches (in the `pred` part) is work which can be reused for the `f` part).

nalimilan commented:
> By specifying the signatures in .md files, it's possible to show only the docstring for the string version in "strings.md", and the general collections version in "collection.md". But unfortunately, the link from replace! choses to point at the string version. In the docs of Documenter, I didn't see a way to address that, someone has an idea? otherwise I will simply remove the link.

I see. Maybe something like ``[`replace`](@ref replace(pred, A, new, [n]))`` would work (adapting the signature)? Else you'll have to ask @mortenpi.

> I don't see that. Assuming there is replace(A, oldnew::Pair...; [count]), how do you propose the version with a predicate to be? (for which old is not necessary anymore).

Mmm, actually I was thinking about the possible ambiguity with the `replace(A, old, new)` form, but that doesn't help with the case you highlighted.

> One possibility is to remove replace(pred, A, new), and to merge merge f and pred for the last version, which can be more powerful:
> replace(pred_f, A), with pred_f returning Nullable for the replacement value. This would allow to implement the pair version in terms of this pred_f (finding whether a pair matches (in the pred part) is work which can be reused for the f part).

Interesting! It makes sense to do everything from a single function indeed. (Regarding the implementation of the pair version, I'm not sure what's your suggestion but you'd have to be very careful that everything is optimized.)

rfourquet commented:
So I implemented your idea, and I like it very much:
1) to have a keyword arg for `n` is way clearer (it should be changed to `count` or something more explicit though)
2) having `old=>new` also is more clear
3) when the keyword is not specified, the speed are on par with the previous version (even faster, I don't understand why yet)
4) I kept the version `replace(pred, A, new)` as it doesn't hurt after all

Compare:
```
julia> replace(A, 2, 3, 8) # old version

julia> replace(A, 2=>3, count=8) # new version
```
PS: your trick to use `[`replace`](@ref replace(pred, A, new, [n]))` for the doc didn't work :(

nalimilan commented on base/array.jl:
```diff
@@ -1987,3 +1987,162 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+"""
+replace!(A, old_new::Pair...; n::Integer=typemax(Int)) = _replace!(A, eltype(A), n, old_new...)
+
+_replace!(A, ::Type{K}, n::Integer, old_new::Pair...) where {K} = _replace!(A, n) do x
```
Can't you just use `eltype(A)` here? Also I don't understand why you need to a separate method than `replace!`.

rfourquet replied:
For a reason I don't understand, if I use directly `eltype(A)` instead of `K` in this method, the performances decrease by maybe an order of magnitude.
I decided to use `_replace!` instead of `replace!` for those helper functions as they are not part of the API. With the `methods` function, they would be discoverable and people may start to use them, making it more difficult to change them. BTW, I wish the same was done for other functions too in Base, as I find that helper methods with the same name as the documented methods tend to pollute the output of `methods`...

nalimilan replied:
> For a reason I don't understand, if I use directly eltype(A) instead of K in this method, the performances decrease by maybe an order of magnitude.

How do you benchmark `replace!`? I suspect wrapping it inside a function makes a difference.

Naming helpers differently is fine, but here it wouldn't make any difference AFAIK.

rfourquet replied:
This comment failed to be sent yesterday:
> Also, this was necessary to introduce a helper method with `n` as a positional arg instead of a keyword arg, as otherwise forwarding arguments with `n=n` for the keyword was killing performances (in particular when the initial call used the default arg).

I benchmarked `replace` with Benchmarktool, but I need to do it more precisely. In the meantime, I implemented a more direct algorithm for `replace!` taking pairs, and it's faster by about 50-100%. For a base function, maybe it's worth it.

>  Naming helpers differently is fine, but here it wouldn't make any difference AFAIK.

I don't see how it doesn't make a difference (or I misunderstand what you mean): if `_replace!(A, ::Type{K}, n::Integer, old_new::Pair...)` was named `replace!`, this method would not be more discoverable but not part of the API...

nalimilan replied:
> Also, this was necessary to introduce a helper method with n as a positional arg instead of a keyword arg, as otherwise forwarding arguments with n=n for the keyword was killing performances (in particular when the initial call used the default arg).

I wonder why that's the case. I would have thought by default the compiler would do something essentially equivalent to what you did (i.e. wrap inside a function which receives positional arguments).

> I benchmarked replace with Benchmarktool, but I need to do it more precisely. In the meantime, I implemented a more direct algorithm for replace! taking pairs, and it's faster by about 50-100%. For a base function, maybe it's worth it.

That would definitely be worth it, but make sure it happens when wrapping inside a function.

> I don't see how it doesn't make a difference (or I misunderstand what you mean): if _replace!(A, ::Type{K}, n::Integer, old_new::Pair...) was named replace!, this method would not be more discoverable but not part of the API...

No, I meant dropping `_replace!(A, ::Type{K}, n::Integer, old_new::Pair...)` completely in favor of `replace!(A, old_new::Pair...; n::Integer=typemax(Int))`. Of course this is assuming that good performance can be attained without the `_replace!` wrapper.

rfourquet replied:
So good news, the `replace[!](A, old_new::Pair...; n::Integer=typemax(Int))` implemented in terms of the "master" method with a closure (over `old_new`) as `prednew` seems now as fast a direct algorithm (provided the closure is declared to be inline). 

> make sure it happens when wrapping inside a function.
as I benchmark `replace`, which is a wrapper arount `_replace`, I guess this is fine?

As for dropping the `_replace!` wrapper, I will do if I manage to not get degraded performances. In the meantime, I added comments explaining why this wrapper is used.

nalimilan commented on base/array.jl:
````diff
@@ -1987,3 +1987,162 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+Set([0, 2, 3])
+```
+"""
+replace!(A, old_new::Pair...; n::Integer=typemax(Int)) = _replace!(A, eltype(A), n, old_new...)
````
Break lines at 92 chars here and elsewhere. Also I'm not a fan of the name `old_new`, but I don't really have better ideas.

rfourquet replied:
OK. And same opinion about `old_new`...

nalimilan commented on base/array.jl:
````diff
@@ -1987,3 +1987,162 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+```jldoctest
+julia> A = [1, 2, 3, 1];
+
+julia> replace!(isodd, A, 0, n=2); a
````
`a` does not exist. Just remove it (I think we discussed this somewhere already).

rfourquet replied:
Oups, I was changing `a` to `A` to match the docstring, but didn't go all the way! After my older comment that I put it this way to make it explicit that `A` is mutated, you didn't insist so I kept it ;-p (changed now).

nalimilan commented on base/set.jl:
```diff
@@ -351,3 +351,27 @@ end
+askey(k, ::AbstractSet) = k
+
+function _replace!(prednew::Callable, A::Union{Associative,AbstractSet}, n::Int)
+    # precondition: n > 0
```
I'd just check this again, that's really cheap anyway.

rfourquet replied:
Sure, I did that for `Abstractarray` and forgot to propagate the change here.

nalimilan commented on doc/src/stdlib/collections.md:
```diff
@@ -133,6 +133,9 @@ Base.collect(::Type, ::Any)
 Base.issubset(::Any, ::Any)
 Base.filter
 Base.filter!
+# this also includes the docstring for replace(A, old, new, n)
```
Outdated comment. And how it is possible that docstrings with a different signature are included?

rfourquet replied:
I have no idea; if I include both signature, the docstring appear twice. This is very brittle, I must have done something wrong, would be nice if a documentation expert could help on this.

nalimilan replied:
Can you post the exact code you tried? That could help somebody more familiar with these issues.

rfourquet replied:
So I cannot reproduce the problem I reported here...

nalimilan commented on test/arrayops.jl:
```diff
@@ -2149,3 +2149,31 @@ Base.:(==)(a::T11053, b::T11053) = a.a == b.a
+    @test replace(s, 1=>4) == Set([2, 3, 4])
+    @test replace!(s, 1=>2) == Set([2, 3])
+
+    @test 0 ∉ replace([1, 2, 3], 1=>0, n=0) # count=0 --> no replacements
```
Just test `== [1, 2, 3]`, that's stricter.

nalimilan commented on test/arrayops.jl:
```diff
@@ -2149,3 +2149,31 @@ Base.:(==)(a::T11053, b::T11053) = a.a == b.a
+    @test a == [1, 4, 3, 1]
+    @test replace(a, 1=>0) == [0, 4, 3, 0]
+    @test replace(a, 1=>0, n=1) == [0, 4, 3, 1]
+    @test replace!(a, 1=>2) == [2, 4, 3, 2]
```
Should test using `===` too.

nalimilan commented:
Cool. I'd rename `n` now too since we won't be able to change it later (keyword argument). Maybe "maxcount"? Should check for consistency with other functions, if any. Too bad for the docstrings, maybe somebody else can help.

nalimilan commented:
Ah, and have you checked the generated code when several pairs are passed? You should be able to check that there are no loops in the LLVM/native code, and/or you could check that a function written by hand with a series of `if x ==` conditions isn't significantly faster.

rfourquet commented:
> Cool. I'd rename n now too since we won't be able to change it later (keyword argument). Maybe "maxcount"? Should check for consistency with other functions, if any. 

`maxcount` sounds good. I will change the name when support for this API is confirmed (we may also discuss a similar change for the method on strings).

I will try  (after today) to check the generated code but am not familiar with this. But at least, like I said, my preliminary test show that it tends to be faster than the previous version, except in the case where a conversion occurs: `replace!(A::Vector{Int}, 1=>0x2)` is significantly slower than  `replace!(A, 1, 0x2)` (I will try to solve this).

rfourquet commented:
I renamed `n` to `count`, which is the name used by python for a similar keyword. The suggested `maxcount` is maybe more clear, if a bit verbose.

rfourquet commented:
I'm pretty happy with the current state of implementation, but I feel this would need more reviews, in particular for the API. Any volunteer?

StefanKarpinski commented:
I think I rather like this API. In light of recent thinking about collections APIs, perhaps @JeffBezanson and @shashi should take a look at this?

tkelman commented on base/array.jl:
```diff
@@ -2119,3 +2119,181 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+replace!(A, old_new::Pair...; count::Integer=typemax(Int)) = _replace!(A, eltype(A), count, old_new)
+
+# we use this wrapper because using directly eltype(A) as the type
+# parameter below for Nullable degrades performances
```
performance

tkelman commented on base/array.jl:
````diff
@@ -2119,3 +2119,181 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+
+# Example
+
+```jldoctest
````
```
# Examples
```jldoctest
```

tkelman commented on base/array.jl:
```diff
@@ -2119,3 +2119,181 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+_replace(prednew::Callable, A, count::Integer) =
+    _replace(prednew, A, clamp(count, typemin(Int), typemax(Int)) % Int)
+
+# we use this _replace wrapper because otherwise the performances are
```
performance is

tkelman commented on base/array.jl:
```diff
@@ -2119,3 +2119,181 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+# we use this _replace wrapper because otherwise the performances are
+# degraded when forwarding the keyword argument
+function _replace!(prednew::Callable, A::AbstractArray, count::Int)
+    count < 0 && throw(DomainError())
```
should now use a value and message

tkelman commented on base/array.jl:
```diff
@@ -2119,3 +2119,181 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+    @inbounds for i in eachindex(A)
+        y = prednew(A[i])
+        if !isnull(y)
+            A[i] = get(y)
```
this could be unsafe_get to avoid checking the null flag twice?

rfourquet replied:
thanks for your precise review!

tkelman commented on base/array.jl:
````diff
@@ -2119,3 +2119,181 @@ julia> symdiff([1,2,3],[3,4,5],[4,5,6])
+value will be used as a replacement for `x`.
+
+# Example
+```
````
```
# Examples
```jldoctest
```

tkelman replied:
this could be a doctest, couldn't it?

rfourquet replied:
yes sorry, fixed

tkelman commented on base/set.jl:
```diff
@@ -422,3 +422,28 @@ end
+askey(k, ::AbstractSet) = k
+
+function _replace!(prednew::Callable, A::Union{Associative,AbstractSet}, count::Int)
+    count < 0 && throw(DomainError())
```
should now pass a value and a message

rfourquet replied:
Done, but I see that some `@noinline` were added by Tim in his PR, should I bother with that here?

tkelman commented on base/set.jl:
```diff
@@ -422,3 +422,28 @@ end
+    for x in A
+        y = prednew(x)
+        if !isnull(y)
+            push!(repl, x=>get(y))
```
could use unsafe_get

rfourquet commented:
> Maybe something like [`replace`](@ref replace(pred, A, new, [n])) 

@nalimilan, I see that I had changed it to the updated version of this, but without the optional keyword, and I discover only now that it works!

rfourquet commented:
I rebased and updated with a (WIP) version to try out with `Union{Some{T}, Void}` instead of `Nullable{T}`. Performance degrades a bit apparently (by about 30 or 35% in the very few examples I tested) . For example `a=[1:100;]; replace(x->Nullable(2x, iseven(x)), a))` takes 165ns when `replace(x->if iseven(x) Some(2x) end, a)` takes 227ns.

nalimilan commented:
Maybe better have two versions of the function for now instead of relying on `Nullable` to unify them? Duplication is limited so that's not the end of the world.

rfourquet commented:
> Maybe better have two versions of the function for now instead of relying on Nullable to unify them?

I think I'd rather pay the 30% overhead price for now, in particular if some optimizations are in the pipeline! It's a new functionality, so no performance will be degraded in user code, and 30% isn't so big.

nalimilan commented:
Why not. But we should leave an issue open so that we don't forget to run the benchmarks again once the optimizations have landed to ensure that approach is indeed fast at that point.

nalimilan commented:
@rfourquet I guess this won't be in 0.7, but maybe we should deprecate `replace(s, pattern, replacement)` in favor of `replace(s, pattern => replacement)` now to avoid ambiguity issues in the future when adding more powerful functions? I also find it more explicit.

StefanKarpinski commented:
I like that idea. It would also allow giving multiple replacements, which can potentially be done more efficiently in a single pass.

rfourquet commented:
That sounds good indeed. I thought we didn't reach a decision yet on this design with pairs, but I still like it. It's true that I didn't plan to get it into 0.7, but mostly because it's not breaking and I didn't want to overload triage, so this could wait for 1.1. But if we go for this design and deprecate the existing `replace`, then given all the review which has gone into this PR already, I would propose to merge it too. I will rebase to get an updated feel of how it looks with your `Union{Some{T}, Void}` PR merged.

rfourquet commented:
Let me recap the API proposed here, if it can help:

1. `replace(A, 1=>2, 2=>3; count=3)` to replace in `A` 1 by 2 and 2 by 3, with at most 3 replacements (`count` available for all methods).
2. `replace(isodd, A, 0)` to replace odd numbers by 0 in `A`
3. `replace(x -> isodd(x) ? 2x : nothing)` to replace odd numbers `x` by `2x`. Here, `2x` can be written as `Some(2x)` (`Some` will be useful when you want to replace a value by `nothing`).

Also included the mutating variant, `replace!`.

nalimilan commented on base/set.jl:
```diff
@@ -474,3 +474,220 @@ end
+
+# we use this wrapper because using directly eltype(A) as the type
+# parameter below for Some degrades performance
+function _replace!(A, ::Type{K}, count::Integer, old_new #=::Pair...=#) where {K}
```
Mark `old_new` with `::Tuple{Vararg{Pair}}`?

rfourquet replied:
Good catch!

nalimilan commented on base/set.jl:
```diff
@@ -474,3 +474,220 @@ end
+## replace/replace! ##
+
+# NOTE: to implement replace! and replace for a new type T, it's enough to define:
+# function _replace!(pred::Callable, A::T, count::Int)
```
Again, I don't really like recommending people to overload an internal function. Is there a way to make this `replace!(pred::Callable, A::T; count::Int)`? BTW, keyword arguments are supposed to be faster now.

rfourquet replied:
Ah right, I didn't re-evaluate this choice in light of the supposedly faster keyword args. I would be willing to benchmark that, but as I'm very short on time now, I will accept your offer to address that later by giving it a lower priority.

nalimilan replied:
Why not go with the simplest solution of defining only the keyword argument version for now, and see later whether performance can be improved?

nalimilan commented on base/set.jl:
````diff
@@ -474,3 +474,220 @@ end
+
+# Examples
+```jldoctest
+julia> replace!(Dict(1=>2, 3=>4)) do kv
````
Could you add a simpler example which doesn't use the `do` syntax, and an example using `Some`?

nalimilan commented on base/set.jl:
```diff
@@ -474,3 +474,220 @@ end
+replace(prednew::Callable, A; count::Integer=typemax(Int)) = _replace!(prednew, copy(A), count)
+
+# Handle ambiguities
+replace!(a::Callable, b::Pair; count::Integer=-1) = throw(MethodError(replace!, a, b))
```
Should be `(a, b)`. Else you get a `MethodError`, but it happens when failing to create the `MethodError`. ;-)

But should these really be errors? It's perfectly valid to have an array of pairs and to replace some pairs with others.

rfourquet replied:
Ah yes right! Those were not tested, just inserted to silence some ambiguities. You are right that what you said is perfectly valid, but this method doesn't have an array as parameter! This is just a silly signature, but Julia needs it!

nalimilan commented on base/set.jl:
```diff
@@ -474,3 +474,220 @@ end
+### _replace! for AbstractDict/AbstractSet
+
+askey(k, ::AbstractDict) = k.first
+askey(k, ::AbstractSet) = k
```
These two convenience functions could interest @andyferris, who tried to unify iteration over sets and dicts.

nalimilan replied:
Maybe that would justify defining `pairs(::AbstractSet)`.

vtjnash replied:
How so? That seems backwards from the problem. The issue here seems to be that `pop!(::AbstractDict, key)` / `delete!` is inconsistent with the way the rest of the way the Associative interface is defined (e.g. it should remove an element, not a key). For consistency with our other indexed containers, these methods should perhaps be renamed `splice!` or `deleteat!`. (later in v1.x we can then re-add `pop!(dict, element)`)

nalimilan replied:
FWIW, `deleteat!` vs `delete!` has been discussed at https://github.com/JuliaLang/julia/pull/20531.

nalimilan commented on base/set.jl:
```diff
@@ -474,3 +474,220 @@ end
+askey(k, ::AbstractDict) = k.first
+askey(k, ::AbstractSet) = k
+
+function _replace_update_hash!(repl::Vector{<:Pair}, x, y::Some)
```
I don't understand why it's called "hash". Maybe "dict" instead?

rfourquet replied:
I just had to quicly find a name, and both `Dict` and `Set` use a hash, but it's a wrong name indeed as this is a helper function for `Union{AbstractDict,AbstractSet}` in general

nalimilan commented on base/set.jl:
```diff
@@ -474,3 +474,220 @@ end
+    repl = Pair{eltype(A),eltype(A)}[]
+    c = 0
+    for x in A
+        c +=_replace_update_hash!(repl, x, prednew(x))
```
Isn't it possible to do the replacement in a single pass?

Missing space after `+=`.

rfourquet replied:
I think it's possible for particular containers, but in general we don't know if mutating the container while iterating is valid (e.g. for a hash-table, there could be a rehash after one insertion, and the rest of the iteration may not visit all non-visited elements. As a first step I wanted something always safe, but it's likely that this can be improved at least for `Set` and `Dict`.

nalimilan commented on base/set.jl:
```diff
@@ -474,3 +474,220 @@ end
+    c = 0
+    for i in eachindex(A)
+        c += _replace_update!(A, i, prednew(A[i]))
+        c == count && break
```
That's probably an overkill optimization, but I wonder whether using a special loop for `count == typemax(Int)` could be worth it in some cases where the function is so simple that SIMD might be enabled without the `break` call. Maybe leave that to future improvements after some benchmarking.

rfourquet replied:
I will add that on my todo list to not forget :)

nalimilan commented on doc/src/stdlib/strings.md:
```diff
@@ -38,7 +38,7 @@ Base.searchindex
 Base.contains(::AbstractString, ::AbstractString)
 Base.reverse(::Union{String,SubString{String}})
-Base.replace
+Base.replace(s::AbstractString, pat, f)
```
This old method is inconsistent with the new ones. Maybe deprecate it in favor of `pat => f`?

rfourquet replied:
Done :)

nalimilan commented on test/sets.jl:
```diff
@@ -344,3 +344,37 @@ end
 end
+
+@testset "replace! & replace" begin
+    maybe1(v, p) = if p Some(v) end
```
The point of using `Some` is to be able to distinguish `nothing`. Can you add a test with `Some(nothing)` to ensure this works?

nalimilan reviewed (commented):
Thanks! If @StefanKarpinski is fine with merging this now, I think we should go ahead. Most of my comments could be addressed later as they don't affect the API, but I think the `AbstractString`-specific method should be deprecated (see below).

rfourquet commented:
I addressed most of comments @nalimilan thanks!
I just kept the deprecation for `replace(::String, ...)` for another PR.
I ended up doing few benchmarks concerning the `_replace!` methods which avoids forwarding keywords arguments, and it seems there is basically no overhead now with that (except 1 allocation, which doens't affect the performance), so I followed your suggestion to get rid of this `_replace!`.

nalimilan reviewed (approved):
Thanks! Looks good, though it needs a NEWS entry.

The deprecation of the other method should also happen quite quickly since we're theoretically past feature freeze.

vtjnash commented on base/set.jl:
```diff
@@ -474,3 +474,249 @@ end
+        c == count && break
+    end
+    for oldnew in repl
+        pop!(A, askey(first(oldnew), A))
```
should this do all deletions before doing all insertions?

rfourquet replied:
I did it this way to avoid 2 passes, and as a result had to put a warning in the docstring. But you are definitely right. It's much more predictable, and the basic benchmarks I just did don't show any overhead :) thanks!

vtjnash replied:
splitting the `oldnew` replacements list into `old` and `new` lists might even help performance? (for non-isbits `Pair{Key, Value}`)

nalimilan commented:
@rfourquet I've tried rebasing via GitHub but I apparently failed. Would you do it, so that the PR can be merged ASAP?

rfourquet commented:
> Would you do it, so that the PR can be merged ASAP?

Sure, but I thought it would need a more clear decision? I put earlier today the triage label on the `replace` PR for strings, I guess the outcome of it can apply here too.

nalimilan commented:
Sure, but there seems to be a broad consensus that it's a positive change, so...

rfourquet commented:
Triage said yes to this API for strings, so I will take it we go with this one too! CI is remarkably all green and there are not yet conflicts, I would say this is a good oppotunity window for someone to click the green button.

nalimilan commented:
@rfourquet Can you remind me why we use `nothing` to mean "no replacement"/"keep existing value"? Couldn't we just require the user function to return the original value in that case? Then `nothing` could mean "drop value".

rfourquet commented:
> Can you remind me why we use nothing to mean "no replacement"/"keep existing value"? 

Because I initially used `Nullable`, where `isnull` means "no replacement", I hadn't thought of alternatives.

> Couldn't we just require the user function to return the original value in that case? Then nothing could mean "drop value".

Sounds like an excellent idea! then it's a bit more than purely replacing values, but after all, `replace` for strings already allows deleting some parts (by providing an empty string as the replacement). I started playing with your idea and this seems to work well.  Unless you like to work on this yourself, I will look at the performance tomorrow and prepare a PR to discuss further this API.

nalimilan commented:
Cool. We don't necessarily need to allow dropping some elements, but if we stop using `nothing` for that purpose we can always allow this later if we want.

rfourquet commented:
My preliminary benchmarks seem to indicate mainly no performance regression :)

>  We don't necessarily need to allow dropping some elements, but if we stop using nothing for that purpose we can always allow this later if we want.

indeed, except when the replacement value is `nothing`.
